### PR TITLE
Allow goto in scripts

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/interactive

**Comment by cston** [comment]
@gafter, @jaredpar

File: src/Compilers/CSharp/Portable/Binder/ScriptInitializerBinder.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly ImmutableArray<LabelSymbol> _labels;
+
+        internal ScriptInitializerBinder(SynthesizedInteractiveInitializerMethod scriptInitializer, CompilationUnitSyntax syntax, Binder next) : base(next)
+        {
+            _labels = GetLabels(scriptInitializer, syntax);
+        }
+
+        protected override ImmutableArray<LabelSymbol> BuildLabels()
+        {
+            return _labels;
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/Binder/ScriptInitializerBinder.cs]: It looks like this was intended to be lazy.  Do we have a reason for not following the pattern?

File: src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs [hunk 2]
```diff
@@ -53,11 +52,14 @@ private void Visit(CSharpSyntaxNode syntax, Binder enclosing)
 
         public override void VisitCompilationUnit(CompilationUnitSyntax node)
         {
-            foreach (MemberDeclarationSyntax member in node.Members)
+            var binder = new ScriptInitializerBinder((SynthesizedInteractiveInitializerMethod)_method, node, _enclosing);
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs]: How do we know we can make this cast unconditionally?

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs [hunk 3]
```diff
@@ -152,42 +152,34 @@ internal struct ProcessedFieldInitializers
...
-                    Binder scriptClassBinder = binderFactory.GetBinder(initializerNode);
-                    Debug.Assert(((ImplicitNamedTypeSymbol)scriptClassBinder.ContainingMemberOrLambda).IsScriptClass);
-
-                    if (firstDebugImports == null)
+                    if (parentBinder == null)
                     {
-                        firstDebugImports = scriptClassBinder.ImportChain;
+                        var syntaxTree = syntaxRef.SyntaxTree;
+                        var binderFactory = compilation.GetBinderFactory(syntaxTree);
+                        var compilationUnit = (CompilationUnitSyntax)syntaxTree.GetRoot();
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: Can we ever get here during speculative binding?  If so, can we still assume the root is a ```CompilationUnitSyntax```?

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs [hunk 4]
```diff
@@ -152,42 +152,34 @@ internal struct ProcessedFieldInitializers
...
 
-                    Binder parentBinder = new ExecutableCodeBinder((CSharpSyntaxNode)syntaxRef.SyntaxTree.GetRoot(), scriptInitializer, scriptClassBinder);
-
+                    var syntax = (CSharpSyntaxNode)syntaxRef.GetSyntax();
                     BoundInitializer boundInitializer;
                     if ((object)fieldSymbol != null)
                     {
                         boundInitializer = BindFieldInitializer(
-                            new LocalScopeBinder(parentBinder).WithAdditionalFlagsAndContainingMemberOrLambda(parentBinder.Flags | BinderFlags.FieldInitializer, fieldSymbol),
+                            parentBinder.WithAdditionalFlagsAndContainingMemberOrLambda(BinderFlags.FieldInitializer, fieldSymbol),
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: Why remove the ```LocalScopeBinder```?  I seem to recall that we used to have problems with code walking up the chain looking for one and not finding any.  This might have been inserted to prevent that.

**Comment by amcasey** [comment]
Can labels hide each other?  If not, is there a test confirming that?

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs (hunk 4 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: The ExecutableCodeBinder will create a ScriptInitializerBinder which is a LocalScopeBinder.

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: I'm not sure if we can get here with something other than a CompilationUnitSyntax, but this change was unnecessary so I've reverted it.

File: src/Compilers/CSharp/Portable/Binder/ScriptInitializerBinder.cs (hunk 1 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/ScriptInitializerBinder.cs]: Changed.

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs [hunk 5]
```diff
@@ -106,7 +107,7 @@ internal struct ProcessedFieldInitializers
                             firstDebugImports = parentBinder.ImportChain;
                         }
 
-                        parentBinder = new LocalScopeBinder(parentBinder).WithAdditionalFlagsAndContainingMemberOrLambda(parentBinder.Flags | BinderFlags.FieldInitializer, fieldSymbol);
+                        parentBinder = new LocalScopeBinder(parentBinder).WithAdditionalFlagsAndContainingMemberOrLambda(BinderFlags.FieldInitializer, fieldSymbol);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: Is this change just a cleanup that removes redundant operation?

Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: Yes, ```WithAdditionalFlagsAndContainingMemberOrLambda``` uses ```this.Flags``` which in this case is ```parentBinder.Flags```.

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs [hunk 6]
```diff
@@ -150,44 +151,44 @@ internal struct ProcessedFieldInitializers
...
+                    var scriptClassBinder = binderFactory.GetBinder(syntax);
+                    Debug.Assert(((NamedTypeSymbol)scriptClassBinder.ContainingMemberOrLambda).IsScriptClass);
 
                     if (firstDebugImports == null)
                     {
                         firstDebugImports = scriptClassBinder.ImportChain;
                     }
 
-                    Binder parentBinder = new ExecutableCodeBinder((CSharpSyntaxNode)syntaxRef.SyntaxTree.GetRoot(), scriptInitializer, scriptClassBinder);
+                    var parentBinder = new ExecutableCodeBinder(syntaxRoot, scriptInitializer, scriptClassBinder).GetBinder(syntaxRoot);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: When does GetBinder call make a difference?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: BTW, this is an illegal use of ```var```. The type is not explicitly spelled out on the right side of the assignment. Please revert.

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: Assuming that LocalScopeBinder propagates them.

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs [hunk 7]
```diff
@@ -150,44 +151,44 @@ internal struct ProcessedFieldInitializers
...
-                    {
-                        // TODO: labels in interactive
-                        var boundStatement = new BoundBadStatement(initializerNode, ImmutableArray<BoundNode>.Empty, true);
-                        boundInitializer = new BoundGlobalStatementInitializer(initializerNode, boundStatement);
-                    }
                     else
                     {
-                        var collisionDetector = new LocalScopeBinder(parentBinder);
-                        boundInitializer = BindGlobalStatement(collisionDetector, scriptInitializer, (StatementSyntax)initializerNode, diagnostics,
+                        var statement = (GlobalStatementSyntax)syntax.Parent;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: Is this a safe cast?

**Comment by cston** [comment]
@amcasey Added hiding tests.

File: src/Compilers/CSharp/Portable/Binder/ScriptInitializerBinder.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        protected override ImmutableArray<LabelSymbol> BuildLabels()
+        {
+            var builder = ArrayBuilder<LabelSymbol>.GetInstance();
+            foreach (var member in _syntax.Members)
+            {
+                if (member.Kind() != SyntaxKind.GlobalStatement)
+                {
+                    continue;
+                }
+                LocalScopeBinder.BuildLabels(_scriptInitializer, ((GlobalStatementSyntax)member).Statement, ref builder);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/ScriptInitializerBinder.cs]: Consider adding an empty line before this statement.

File: src/Compilers/CSharp/Portable/Binder/ScriptInitializerBinder.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+            }
+            return builder.ToImmutableAndFree();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/ScriptInitializerBinder.cs]: Consider adding an empty line before this statement.

File: src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs [hunk 10]
```diff
@@ -30,13 +28,13 @@ internal static MethodBodySemanticModel Create(CSharpCompilation compilation, Me
...
+        /// Creates a SemanticModel for a GlobalStatementSyntax.
         /// </summary>
-        internal static MethodBodySemanticModel Create(CSharpCompilation compilation, MethodSymbol owner, Binder rootBinder, ArrowExpressionClauseSyntax syntax)
+        internal static MethodBodySemanticModel Create(CSharpCompilation compilation, SynthesizedInteractiveInitializerMethod scriptInitializer, Binder rootBinder, GlobalStatementSyntax syntax)
         {
-            Binder binder = new ExecutableCodeBinder(syntax, owner, rootBinder);
-            return new MethodBodySemanticModel(compilation, owner, binder, syntax);
+            var compilationUnit = (CompilationUnitSyntax)syntax.Parent;
+            var binder = new ExecutableCodeBinder(compilationUnit, scriptInitializer, rootBinder);
+            return new MethodBodySemanticModel(compilation, scriptInitializer, binder.GetBinder(compilationUnit), syntax);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs]: Are we creating MethodBodySemanticModel for each global statement? Is this the right thing to do? Shouldn't all global statements share the same scope, local symbols, etc.?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs]: Which means they should share the binder?

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 11]
```diff
@@ -982,12 +982,12 @@ private MemberSemanticModel CreateMemberModel(CSharpSyntaxNode node)
                         // TODO (tomat): handle misplaced global statements
                         if (node.Parent.Kind() == SyntaxKind.CompilationUnit)
                         {
-                            var scriptConstructor = this.Compilation.ScriptClass.InstanceConstructors.First();
+                            var scriptInitializer = this.Compilation.ScriptClass.GetScriptInitializer();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs]: Please spell out the type.

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs (hunk 6 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: GetBinder will return a ScriptInitializerBinder since it is called with the CompilationUnitSyntax. See LocalBinderFactory.VisitCompilationUnit.

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 12]
```diff
@@ -982,12 +982,12 @@ private MemberSemanticModel CreateMemberModel(CSharpSyntaxNode node)
                         // TODO (tomat): handle misplaced global statements
                         if (node.Parent.Kind() == SyntaxKind.CompilationUnit)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs]: It feels that the model should be created for all global statements in the same compilation unit at once, rather than for each statement separately.

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs (hunk 7 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: I believe so but that change is no longer needed. Reverted.

File: src/Compilers/Core/Portable/CodeGen/ILBuilder.cs [hunk 13]
```diff
@@ -70,7 +70,7 @@ internal ILBuilder(ITokenDeferral module, LocalSlotManager localSlotManager, Opt
 
             leaderBlock = _currentBlock = _scopeManager.CreateBlock(this);
 
-            _labelInfos = new SmallDictionary<object, LabelInfo>(ReferenceEqualityComparer.Instance);
+            _labelInfos = new SmallDictionary<object, LabelInfo>();
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeGen/ILBuilder.cs]: Is there an intent to change behavior here? If not, please revert the change.

File: src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs]: Consider adding a comment.

File: src/Compilers/Core/Portable/CodeGen/ILBuilder.cs (hunk 13 above)
Comment (cston) [inline src/Compilers/Core/Portable/CodeGen/ILBuilder.cs]: The behavior change is intentional. Label declarations and references are distinct instances when binding global statements. FWIW, ```PreciseAbstractFlowPass._labels``` uses default EqualityComparer.

**Comment by AlekseyTs** [comment]
We should have SemanticModel tests for global statements. For example, we need to verify that GetSymbolInfo/GetDeclaredSymbol returns consistent information across multiple global statements (for locals, labels, etc.)

File: src/Compilers/Core/Portable/CodeGen/ILBuilder.cs (hunk 13 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeGen/ILBuilder.cs]: I am not sure if we are dealing with symbols here.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeGen/ILBuilder.cs]: Also, I have no idea what does "Label declarations and references are distinct instances" mean. This doesn't make any sense to me. Let's talk offline.

File: src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs (hunk 10 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs]: Changed to share labels across all global statements.

File: src/Compilers/Core/Portable/CodeGen/ILBuilder.cs (hunk 13 above)
Comment (cston) [inline src/Compilers/Core/Portable/CodeGen/ILBuilder.cs]: Reverted this change now that labels are shared across global statements.

**Comment by cston** [comment]
@AlekseyTs Added SemanticModel tests.

**Comment by cston** [comment]
@dotnet-bot retest this please

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs [hunk 14]
```diff
@@ -53,12 +53,14 @@ internal void LookupExtensionMethods(LookupResult result, string name, int arity
...
             Debug.Assert(options.AreValid());
 
             // don't create diagnosis instances unless lookup fails
-            this.LookupSymbolsInternal(result, name, arity, basesBeingResolved, options, diagnose: false, useSiteDiagnostics: ref useSiteDiagnostics);
+            var binder = this.LookupSymbolsInternal(result, name, arity, basesBeingResolved, options, diagnose: false, useSiteDiagnostics: ref useSiteDiagnostics);
+            Debug.Assert((binder != null) || result.IsClear);
+
             if (result.Kind != LookupResultKind.Viable && result.Kind != LookupResultKind.Empty)
             {
                 result.Clear();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: Should this invalidate the ```binder```?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: If we expect the next call to return the same binder, consider adding an assert to that regard.

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs (hunk 14 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: Added assert.

**Comment by cston** [comment]
Merged in #6076

**Comment by cston** [comment]
Merged in #6076
